.github/workflows/pre-commit.yaml

```diff
@@ -59,33 +59,38 @@ jobs:
 
       - name: Config Terraform plugin cache
         if: steps.changes.outputs.src== 'true'
-        run: mkdir --parents ${{ env.TERRAFORM_DOCS_VERSION }}
+        run: mkdir --parents ${{ env.TF_PLUGIN_CACHE_DIR }}
 
-      - name: Cache Terraform
-        uses: actions/cache@v3
+      - name: Restore Terraform Cache
+        id: restore-cache
+        uses: actions/cache/restore@v4
         if: steps.changes.outputs.src== 'true'
         with:
-          path: ${{ env.TERRAFORM_DOCS_VERSION }}
-          key: ${{ runner.os }}-terraform-${{ hashFiles('**/.terraform.lock.hcl') }}
-          restore-keys: ${{ runner.os }}-terraform-
+          # restore latest matching cache to the tf cache dir
+          # matches exact branch and versions hash, then any cache from the same branch, then any cache from any branch
+          path: ${{ env.TF_PLUGIN_CACHE_DIR }}
+          key: ${{ runner.os }}-terraform-${{ github.head_ref }}-${{ hashFiles('**/versions.tf') }}
+          restore-keys: |
+            ${{ runner.os }}-terraform-${{ github.head_ref }}
+            ${{ runner.os }}-terraform-
```

Comment:
Looks like `restore-keys` is used when key misses.  Is it ok to restore them from any available cache?

An example scenario: There's a cache for Terraform provider plugins from v1.0.0, and a workflow runs that needs v0.9.0 (an older version). 
1. Key misses because cache doesn't exist for this particular version.
2. `terraform-${{ github.head_ref }}` this misses because it's a new branch.
3. Restore from `terraform-`. 

Any concerns here? Or does Terraform handle this kind of this well?

Reply:
Terraform seems to handle it alright, I think it should be able to keep a subdir for v1.0.0 and v0.9.0. The hope is that the _other_ modules that were in the cache are still valid and can be used. 

The fallback is a hope that we can inherit _some_ of the modules from either the same branch or even "most recent cache from any branch". The worst cases I expect are to have to download the missed modules, and some cache bloat if there are multiple package versions like that.